### PR TITLE
WT-6349 Don't perform on disk key removal as part of checkpoint

### PR DESCRIPTION
**Pull request by kommiharibabu**
On disk key removal needs to clear all the history store content
for the key and it might take quite a bit of time when there are
multiple keys that are needs to be removed. To avoid checkpoint
waiting for all these cleanup, better to do this as part of eviction.

### CODE REVIEW

**Comment by agorrod** [comment]
@kommiharibabu would you mind running a MongoDB patch build with this change and making sure it looks OK?

**Comment by kommiharibabu** [comment]
The existing on-disk value is written to disk instead of removing them during reconciliation. But one problem in this approach is that they will not be clean up until eviction doesn't evict this page. we shouldn't remove the key, if we do that, the history store must be traversed to clean up.

**Comment by quchenhao** [comment]
> The existing on-disk value is written to disk instead of removing them during reconciliation. But one problem in this approach is that they will not be clean up until eviction doesn't evict this page. we shouldn't remove the key, if we do that, the history store must be traversed to clean up.

That doesn't sounds correct to me. If we have a mixed mode deletion, we should remove that in the checkpoint as well. Otherwise the checkpoint is inconsistent.

**Comment by kommiharibabu** [comment]
> That doesn't sound correct to me. If we have a mixed mode deletion, we should remove that in the checkpoint as well. Otherwise the checkpoint is inconsistent.

The key will be removed in mixed mode scenarios. In mixed mode scenario, the `upd` will not be NULL during reconciliation. The only case which we are avoiding here is the timestamp delete operation that is written to disk which wasn't globally visible during reconciliation.

File: src/reconcile/rec_row.c [hunk 1]
```diff
@@ -770,9 +770,11 @@ __wt_rec_row_leaf(
 
         /*
          * If we reconcile an on disk key with a globally visible stop time point and there are no
-         * new updates for that key, skip writing that key.
+         * new updates for that key, skip writing that key. Don't do this additional cleanup work if
+         * a checkpoint is doing the work - we need that to be as streamlined as possible.
          */
-        if (upd == NULL && __wt_txn_tw_stop_visible_all(session, &tw))
+        if (upd == NULL && !F_ISSET(r, WT_REC_CHECKPOINT) &&
+          __wt_txn_tw_stop_visible_all(session, &tw))
```
Comment (tetsuo-cpp) [inline src/reconcile/rec_row.c:777]: I think I understand this.
By definition, we should never read this OR the associated history store content.
But we're not removing it from the disk image yet (and temporarily wasting space) to not block the checkpoint.

Is that right @kommiharibabu?

Comment (kommiharibabu) [reply src/reconcile/rec_row.c:777]: Yes, your understanding is correct. We are not removing any on-disk tombstones that are globally visible as part of the checkpoint.

**Comment by kommiharibabu** [comment]
The evergreen failure is WT-6263.
I ran a patch build and one test failed because of validation errors.
https://evergreen.mongodb.com/task/mongodb_mongo_v4.4_enterprise_rhel_62_64_bit_display_rollback_fuzzer_unclean_shutdowns_patch_903bcf6579085c135f9506933d8f671f54255eb0_5ed471c261837d66b7fb4be1_20_06_01_03_11_54##comparehashes=903bcf6579085c135f9506933d8f671f54255eb0&threads=maxonly

I will check for the reason and update it.

**Comment by tetsuo-cpp** [comment]
> The evergreen failure is WT-6263.
> I ran a patch build and one test failed because of validation errors.
> https://evergreen.mongodb.com/task/mongodb_mongo_v4.4_enterprise_rhel_62_64_bit_display_rollback_fuzzer_unclean_shutdowns_patch_903bcf6579085c135f9506933d8f671f54255eb0_5ed471c261837d66b7fb4be1_20_06_01_03_11_54##comparehashes=903bcf6579085c135f9506933d8f671f54255eb0&threads=maxonly
> 
> I will check for the reason and update it.

Just FYI, there was some fallout from WT-5769 which I'm addressing now. Not sure if that's what is behind that failure but just something to keep in mind.

**Comment by kommiharibabu** [comment]
WT-5769 is not the reason behind the failure. I didn't include those changes as part of the MongoDB patch build. The failures are may be from an existing open issue, need to confirm.

**Comment by kommiharibabu** [comment]
The issue seems to be related to BF-16779. May be easy to reproduce with this change. I didn't yet reproduce the issue locally.

**Comment by kommiharibabu** [comment]
@agorrod @tetsuo-cpp @quchenhao The patch build failures are not related to the changes that are done as part of this ticket, I created an another ticket WT-6395 to handle the other problem.
I will merge this pull request. Does anyone have any concerns to merge this change?

**Comment by agorrod** [comment]
@kommiharibabu if we merge this now, will it mean we see a lot of test failures in Evergreen due to WT-6395?

**Comment by kommiharibabu** [comment]
One test will fail due to the open issue of WT-6395, but it can multiply due to many variant hosts.

**Comment by agorrod** [comment]
Sadly, I think we need to hold off on merging this change until we can get a fix for WT-6395

**Comment by agorrod** [comment]
@kommiharibabu could you update to the latest develop and kick off another round of testing on this once WT-6395 is in MongoDB.
If testing looks good I'll be very happy for this change to go in.

**Comment by agorrod** [comment]
@kommiharibabu you have bad PR testing karma. The PR failure is a known failure. LGTM - merge when you are ready.

**Comment by kommiharibabu** [comment]
@agorrod I didn't include WT-6395 when I ran the stress tests yesterday, but I include them in MongoDB patch build. I will run the stress tests tonight and if everything is good. Will merge it tomorrow morning.

**Comment by agorrod** [comment]
SGTM

**Comment by agorrod** [comment]
@kommiharibabu is this ready to be merged now?

**Comment by kommiharibabu** [comment]
@agorrod Yes, it can be merged now. I wasn't able to reproduce the failure locally and also the tests are passed in evergreen also. As the failure doesn't have any relation to the changes, so it can be merged now.

**Pull request merged**